fix validate_path_is_safe function on windows

serena-ruan opened the pull request:
<!-- 🚨 We recommend pull requests be filed from a non-master branch on a repository fork (e.g. <username>:fix-xxx). 🚨 -->

## Related Issues/PRs

<!--
Please reference any related feature requests, issues, or PRs here. For example, `#123`. To automatically close the referenced issues when this PR is merged, please use a closing keyword (close, fix, or resolve). For example, `Resolve #123`. See https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue for more information.
-->
Update validate_path_is_safe function to work properly on windows machine.

## What changes are proposed in this pull request?

(Please fill in changes proposed in this fix)

## How is this patch tested?

<!--
If you're unsure about what to test, where to add tests, or how to run tests, please feel free to ask.
-->

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

<!--
Please describe how you confirmed the proposed feature/bug-fix/change works here. For example, if you fixed an MLflow client API, you could attach the code that didn't work prior to the fix but works now, or if you added a new feature on MLflow UI, you could attach a video that demonstrates the feature.
-->

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

## Release Notes

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

serena-ruan committed:
fix validate_path_is_safe on windows

Signed-off-by: Serena Ruan <serena.rxy@gmail.com>

serena-ruan committed:
update tests

Signed-off-by: Serena Ruan <serena.rxy@gmail.com>

serena-ruan committed:
update test

Signed-off-by: Serena Ruan <serena.rxy@gmail.com>

mlflow-automation commented:
<!-- documentation preview -->


Documentation preview for 5e3104e445724fbb46ac2b09b5100983960e6106 will be available [here](https://output.circle-artifacts.com/output/job/ff18c7fa-6dc9-48d7-ac69-e26d979a0eff/artifacts/0/docs/build/html/index.html) when [this CircleCI job](https://circleci.com/gh/mlflow/mlflow/55779) completes successfully.

<details>
<summary>More info</summary>

- Ignore this comment if this PR does not change the documentation.
- It takes a few minutes for the preview to be available.
- The preview is updated when a new commit is pushed to this PR.
- This comment was created by https://github.com/mlflow/mlflow/actions/runs/5483997611.

</details>

harupy commented on tests/tracking/test_rest_tracking.py:
```diff
@@ -544,6 +559,40 @@ def test_validate_path_is_safe_bad(path):
 
 
+@pytest.mark.skipif(not is_windows(), reason="This test only passes on Windows")
+@pytest.mark.parametrize(
+    "path",
+    [
+        "../path",
+        "../../path",
+        "./../path",
+        "path/../to/file",
+        "path/../../to/file",
+        "..\\path",
+        "..\\..\\path",
+        ".\\..\\path",
+        "path\\..\\to\\file",
+        "path\\..\\..\\to\\file",
+        # Drive-relative absolute paths
+        "C:\\path",
+        "C:/path",
+        "C:\\path\to\file",
+        "C:\\path/to/file",
+        "C:\\path\\..\\to\\file",
+        "C:/path/../to/file"
+        # UNC(Universal Naming Convention) paths
+        "\\\\path\\to\\file",
+        "\\\\path/to/file",
+        "\\\\.\\C:\\path\\to\\file",
+        "\\\\?\\C:\\path\\to\\file",
+        "\\\\?\\UNC/path/to/file",
+    ],
+)
```
/ai rewrite this using raw string and chage `\\` to `\`

harupy commented on tests/tracking/test_rest_tracking.py:
```diff
@@ -528,6 +528,21 @@ def test_validate_path_is_safe_good(path):
+    "path",
+    [
+        # relative path from current directory of C: drive
+        "C:path",
+        "C:path/",
+        "C:path/to/file",
```
Is `C:../path/to/file"` bad?

serena-ruan replied:
emmm I don't think that's a valid path? 
<img width="666" alt="image" src="https://github.com/mlflow/mlflow/assets/82044803/7c203e0e-a8c6-44cd-be3e-c7370ec0f326">

serena-ruan replied:
But if providing this path, it will be identified as a relative path 🤔

harupy replied:
good relative path or bad relative path?

serena-ruan replied:
good relative path I think, it doesn't meet any of the three conditions to throw an exception

BenWilson2 commented on tests/tracking/test_rest_tracking.py:
```diff
@@ -544,6 +559,40 @@ def test_validate_path_is_safe_bad(path):
+@pytest.mark.parametrize(
+    "path",
+    [
+        "../path",
```
some additional Windows-specific attack vectors:

```

    r'../etc/passwd',
    r'..\Windows\System32\',
    r'C:\Windows\System32\',
    r'/etc/passwd',
    r'::Windows\System32',
    r'..\..\..\..\Windows\System32\',
    r'../Windows/System32',
    r'.../...//',
    r'....\\',
    r'\\?\C:\Windows\System32\',
    r'\\.\C:\Windows\System32\',
    r'\\UNC\Server\Share\',
    r'\\Server\Share\folder\',
    r'\\127.0.0.1\c$\Windows\',
    r'\\localhost\c$\Windows\',
    r'\\smbserver\share\path\',
    r'..\\?\C:\Windows\System32\',
    r'file:///C:/Windows/System32/',
    r'C:/Windows/../Windows/System32/',
    r'C:\Windows\..\Windows\System32\',
    r'/Windows/no/such/path',
    r'../../../../../../../../../../../../Windows/System32',
    r'../../../../../../../../../../../../etc/passwd',
    r'../../../../../../../../../../../../var/www/html/index.html',
    r'../../../../../../../../../../../../usr/local/etc/openvpn/server.conf',
    r'../../../../../../../../../../../../Program Files (x86)',
    r'/../../../../../../../../../../../../Windows/System32',
    r'file:///etc/passwd',
    r'file:///d:/windows/repair/sam',
    r'file:///proc/version',
    r'file:///inetpub/wwwroot/global.asa',
    r'file://c:/boot.ini',
    r'/file://etc/passwd',
    r'/file:c:/passwd',
    r'/file://d:/windows/win.ini',
    r'/file://./windows/win.ini',
    r'/file://../windows/win.ini',
    r'/Windows\../etc/passwd',
    r'/Windows\..\Windows\System32\',
    r'/etc/passwd%00.jpg',
    r'/etc/passwd%00.html',
    r'/etc/passwd%00.txt',
    r'/etc/passwd%00.php',
    r'/etc/passwd%00.asp',
    r'/Windows\..\Windows\System32\cmd.exe',
    r'/Windows\..\Windows\System32\msconfig.exe',
    r'/Windows\..\Windows\System32\regedit.exe',
    r'/Windows\..\Windows\System32\taskmgr.exe',
    r'/Windows\..\Windows\System32\control.exe',
    r'/Windows\..\Windows\System32\services.msc',
    r'/Windows\..\Windows\System32\diskmgmt.msc',
    r'/Windows\..\Windows\System32\eventvwr.msc',
```

serena-ruan replied:
<img width="738" alt="image" src="https://github.com/mlflow/mlflow/assets/82044803/619db23e-2bf1-420f-9402-02af20a7acd7">

I thought paths beginning with /xxx is always considered as a relative path on windows right? So it should be fine on windows? like "/path", it should be a good path.

harupy replied:
We can't believe what ChatGPT says.

serena-ruan replied:
I'm just curious why /etc/passwd is not a good path in windows if it's a relative path?

harupy replied:
If it's a relative path, I think it's a good path.

harupy replied:
On windows, is it possible to join `/etc/passwd` and another path?

serena-ruan replied:
I think so, path joining is regardless of relative or absolute path?

harupy commented on mlflow/server/handlers.py:
```diff
@@ -530,7 +530,7 @@ def validate_path_is_safe(path):
         any((s in path) for s in _OS_ALT_SEPS)
         or ".." in path.split(posixpath.sep)
-        or posixpath.isabs(path)
+        or pathlib.Path(path).is_absolute()
```
btw do we need to support windows paths?

harupy replied:
or is the challenge posix path detection?

serena-ruan replied:
Yes, on windows machine 
<img width="238" alt="image" src="https://github.com/mlflow/mlflow/assets/82044803/5a3f5607-4406-4579-a635-23e0373b90e7">

serena-ruan committed:
add more paths

Signed-off-by: Serena Ruan <serena.rxy@gmail.com>

serena-ruan committed:
Merge branch 'master' into lfi

Signed-off-by: Serena Ruan <serena.rxy@gmail.com>

serena-ruan commented:
Tests on windows machine:
Good paths:
<img width="1064" alt="image" src="https://github.com/mlflow/mlflow/assets/82044803/cf52ce90-f5f4-4e83-aa8f-c53cee745781">

serena-ruan committed:
update tests

Signed-off-by: Serena Ruan <serena.rxy@gmail.com>

serena-ruan committed:
only call local_file_uri_to_path when it is file uri

Signed-off-by: Serena Ruan <serena.rxy@gmail.com>

adeadfed commented:
> Tests on windows machine: Good paths: <img alt="image" width="1064" src="https://user-images.githubusercontent.com/82044803/252841829-cf52ce90-f5f4-4e83-aa8f-c53cee745781.png">

Hey @serena-ruan ! Was just passing by this pull request. It seems that the issue is a little bit complicated for Windows. `pathlib.Path` will default to a WindowsPath constructor even if called with a POSIX path on Windows, e.g. `pathlib.Path('/some/posix/path') --> WindowsPath('/some/posix/path') --> WindowsPath('/some/posix/path').is_absolute() == False`

While this might look okay for sole purpose of windows path validation, Python still supports POSIX paths on Windows, so we still have a directory traversal:

 ```
 >>> import pathlib
>>> pathlib.Path('/test').is_absolute()
False
>>> pathlib.Path('/').is_absolute()
False
>>> pathlib.Path('/Windows/System32/drivers/etc/hosts').is_absolute()
False
>>> open('/Windows/System32/drivers/etc/hosts', 'r').readlines()
['ï»¿# Copyright (c) 1993-2009 Microsoft Corp.\n', ... , '# End of section\n']
```

adeadfed commented:
> > > Tests on windows machine: Good paths: <img alt="image" width="1064" src="https://user-images.githubusercontent.com/82044803/252841829-cf52ce90-f5f4-4e83-aa8f-c53cee745781.png">
> > 
> > 
> > Hey @serena-ruan ! Was just passing by this pull request. It seems that the issue is a little bit complicated for Windows. `pathlib.Path` will default to a WindowsPath constructor even if called with a POSIX path on Windows, e.g. `pathlib.Path('/some/posix/path') --> WindowsPath('/some/posix/path') --> WindowsPath('/some/posix/path').is_absolute() == False`
> > While this might look okay for sole purpose of windows path validation, Python still supports POSIX paths on Windows, so we still have a directory traversal:
> > ```
> > >>> import pathlib
> > >>> pathlib.Path('/test').is_absolute()
> > False
> > >>> pathlib.Path('/').is_absolute()
> > False
> > >>> pathlib.Path('/Windows/System32/drivers/etc/hosts').is_absolute()
> > False
> > >>> open('/Windows/System32/drivers/etc/hosts', 'r').readlines()
> > ['ï»¿# Copyright (c) 1993-2009 Microsoft Corp.\n', ... , '# End of section\n']
> > ```
> 
> Hi @adeadfed, thanks so much for this helpful information! Since you're an expert on this, would you be willing to add some test cases in your mind to the test suites so that we could better validate our approach? 😃

At quick glance, the list of paths in your tests is exhaustive imo. The inconsistent behavior of pathlib should be resolved like this:

```
def validate_path_is_safe(path):
    """
    Validates that the specified path is safe to join with a trusted prefix. This is a security
    measure to prevent path traversal attacks.
    """
    if is_file_uri(path):
        path = local_file_uri_to_path(path)
    if (
        any((s in path) for s in _OS_ALT_SEPS)
        or ".." in path.split(os.sep)
        or pathlib.PureWindowsPath(path).is_absolute()
        or pathlib.PurePosixPath(path).is_absolute()
    ):
        raise MlflowException(f"Invalid path: {path}", error_code=INVALID_PARAMETER_VALUE)
```

Though I am wondering if `or ".." in path.split(os.sep)` check can be bypassed now or not. I guess we should see that from the output of the tests :) Hope that helps!

serena-ruan committed:
update as comments

Signed-off-by: Serena Ruan <serena.rxy@gmail.com>

serena-ruan commented:
> > > > Tests on windows machine: Good paths: <img alt="image" width="1064" src="https://user-images.githubusercontent.com/82044803/252841829-cf52ce90-f5f4-4e83-aa8f-c53cee745781.png">
> > > 
> > > 
> > > Hey @serena-ruan ! Was just passing by this pull request. It seems that the issue is a little bit complicated for Windows. `pathlib.Path` will default to a WindowsPath constructor even if called with a POSIX path on Windows, e.g. `pathlib.Path('/some/posix/path') --> WindowsPath('/some/posix/path') --> WindowsPath('/some/posix/path').is_absolute() == False`
> > > While this might look okay for sole purpose of windows path validation, Python still supports POSIX paths on Windows, so we still have a directory traversal:
> > > ```
> > > >>> import pathlib
> > > >>> pathlib.Path('/test').is_absolute()
> > > False
> > > >>> pathlib.Path('/').is_absolute()
> > > False
> > > >>> pathlib.Path('/Windows/System32/drivers/etc/hosts').is_absolute()
> > > False
> > > >>> open('/Windows/System32/drivers/etc/hosts', 'r').readlines()
> > > ['ï»¿# Copyright (c) 1993-2009 Microsoft Corp.\n', ... , '# End of section\n']
> > > ```
> > 
> > 
> > Hi @adeadfed, thanks so much for this helpful information! Since you're an expert on this, would you be willing to add some test cases in your mind to the test suites so that we could better validate our approach? 😃
> 
> At quick glance, the list of paths in your tests is exhaustive imo. The inconsistent behavior of pathlib should be resolved like this:
> 
> ```
> def validate_path_is_safe(path):
>     """
>     Validates that the specified path is safe to join with a trusted prefix. This is a security
>     measure to prevent path traversal attacks.
>     """
>     if is_file_uri(path):
>         path = local_file_uri_to_path(path)
>     if (
>         any((s in path) for s in _OS_ALT_SEPS)
>         or ".." in path.split(os.sep)
>         or pathlib.PureWindowsPath(path).is_absolute()
>         or pathlib.PurePosixPath(path).is_absolute()
>     ):
>         raise MlflowException(f"Invalid path: {path}", error_code=INVALID_PARAMETER_VALUE)
> ```
> 
> Though I am wondering if `or ".." in path.split(os.sep)` check can be bypassed now or not. I guess we should see that from the output of the tests :) Hope that helps!

Thanks! Updated and let's see how the tests run :)

adeadfed commented:
Well, judging by the test output, pathlib now works fine. All failed test cases under the `test_validate_path_is_safe_windows_good` are false positives and will be indeed treated by Python as absolute paths, so they should better be moved to the `test_validate_path_is_safe_windows_bad` list of paths. However, there is indeed a problem with `'..' in path` check because it is now split by `os.sep == \`, not by `posixpath.sep == /`. Is it worth to revert this change? Also, do these tests also run on a linux box? It would probably be good to check if nothing has been broken there as well.

serena-ruan commented:
> Well, judging by the test output, pathlib now works fine. All failed test cases under the `test_validate_path_is_safe_windows_good` are false positives and will be indeed treated by Python as absolute paths, so they should better be moved to the `test_validate_path_is_safe_windows_bad` list of paths. However, there is indeed a problem with `'..' in path` check because it is now split by `os.sep == \`, not by `posixpath.sep == /`. Is it worth to revert this change? Also, do these tests also run on a linux box? It would probably be good to check if nothing has been broken there as well.

Since we don't allow '\' in the path, maybe we just explicitly say `.split('/')`  instead of `.split(os.sep)`? This test is only run on windows because the same path might be disallowed on windows but allowed on linux right? `test_validate_path_is_safe_bad` only runs on linux, with ```
"/path",
        "../path",
        "../../path",
        "./../path",
        "path/../to/file",
        "path/../../to/file",
        "/etc/passwd",
        "/etc/passwd%00.jpg",
        "/etc/passwd%00.html",
        "/etc/passwd%00.txt",
        "/etc/passwd%00.php",
        "/etc/passwd%00.asp",
```

serena-ruan committed:
update tests and split path by /

Signed-off-by: Serena Ruan <serena.rxy@gmail.com>

serena-ruan committed:
add docstring for valid path

Signed-off-by: Serena Ruan <serena.rxy@gmail.com>

harupy reviewed (approved):
LGTM!

harupy commented:
@adeadfed Thanks a lot for the help :)

adeadfed commented:
Hey @harupy & @serena-ruan ! Was a pleasure to work with you to make the MLFlow more secure! I've just one question left to you: would you be willing to file a Github security advisory & validate the corresponsing reports on the huntr.dev platform for #8999 & #9053 now that they have been merged into the main branch? I would greatly appreciate that! Stay safe!

serena-ruan commented:
> Hey @harupy & @serena-ruan ! Was a pleasure to work with you to make the MLFlow more secure! I've just one question left to you: would you be willing to file a Github security advisory & validate the corresponsing reports on the huntr.dev platform for #8999 & #9053 now that they have been merged into the main branch? I would greatly appreciate that! Stay safe!

Github security advisory sounds like a good plan! @harupy WDYT? I'll go to huntr.dev and mark your reported issues as resolved! Thanks so much 😃
